File: java-practice/src/main/java/solutions/Ex03Sort.java
package solutions;

public class Ex03Sort {

    public static int[] quickSort(int[] arr) {
        int[] result = arr.clone();
        quickSort(result, 0, result.length - 1);
        return result;
    }

    private static void quickSort(int[] arr, int lo, int hi) {
        if (lo >= hi) return;
        int pivot = arr[lo + (hi - lo) / 2];
        int i = lo, j = hi;
        while (i <= j) {
            while (arr[i] < pivot) i++;
            while (arr[j] > pivot) j--;
            if (i <= j) {
                int tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp;
                i++; j--;
            }
        }
        quickSort(arr, lo, j);
        quickSort(arr, i, hi);
    }

    public static int[] mergeSort(int[] arr) {
        int[] result = arr.clone();
        mergeSort(result, 0, result.length - 1);
        return result;
    }

    private static void mergeSort(int[] arr, int lo, int hi) {
        if (lo >= hi) return;
        int mid = lo + (hi - lo) / 2;
        mergeSort(arr, lo, mid);
        mergeSort(arr, mid + 1, hi);
        merge(arr, lo, mid, hi);
    }

    private static void merge(int[] arr, int lo, int mid, int hi) {
        int[] tmp = new int[hi - lo + 1];
        int i = lo, j = mid + 1, k = 0;
        while (i <= mid && j <= hi) {
            tmp[k++] = arr[i] <= arr[j] ? arr[i++] : arr[j++];
        }
        while (i <= mid) tmp[k++] = arr[i++];
        while (j <= hi) tmp[k++] = arr[j++];
        System.arraycopy(tmp, 0, arr, lo, tmp.length);
    }

    public static int[] heapSort(int[] arr) {
        int[] result = arr.clone();
        int n = result.length;
        for (int i = n / 2 - 1; i >= 0; i--) heapify(result, n, i);
        for (int i = n - 1; i > 0; i--) {
            int tmp = result[0]; result[0] = result[i]; result[i] = tmp;
            heapify(result, i, 0);
        }
        return result;
    }

    private static void heapify(int[] arr, int n, int i) {
        int largest = i;
        int left = 2 * i + 1, right = 2 * i + 2;
        if (left < n && arr[left] > arr[largest]) largest = left;
        if (right < n && arr[right] > arr[largest]) largest = right;
        if (largest != i) {
            int tmp = arr[i]; arr[i] = arr[largest]; arr[largest] = tmp;
            heapify(arr, n, largest);
        }
    }
}

File: go-practice/solutions/ex03_sort.go
package solutions

// QuickSort 快速排序（原地）
func QuickSort(arr []int) {
	quickSort(arr, 0, len(arr)-1)
}

func quickSort(arr []int, lo, hi int) {
	if lo >= hi {
		return
	}
	pivot := arr[lo+(hi-lo)/2]
	i, j := lo, hi
	for i <= j {
		for arr[i] < pivot {
			i++
		}
		for arr[j] > pivot {
			j--
		}
		if i <= j {
			arr[i], arr[j] = arr[j], arr[i]
			i++
			j--
		}
	}
	quickSort(arr, lo, j)
	quickSort(arr, i, hi)
}

// MergeSort 归并排序（返回新切片）
func MergeSort(arr []int) []int {
	result := make([]int, len(arr))
	copy(result, arr)
	mergeSort(result, 0, len(result)-1)
	return result
}

func mergeSort(arr []int, lo, hi int) {
	if lo >= hi {
		return
	}
	mid := lo + (hi-lo)/2
	mergeSort(arr, lo, mid)
	mergeSort(arr, mid+1, hi)
	merge(arr, lo, mid, hi)
}

func merge(arr []int, lo, mid, hi int) {
	tmp := make([]int, hi-lo+1)
	i, j, k := lo, mid+1, 0
	for i <= mid && j <= hi {
		if arr[i] <= arr[j] {
			tmp[k] = arr[i]
			i++
		} else {
			tmp[k] = arr[j]
			j++
		}
		k++
	}
	for i <= mid {
		tmp[k] = arr[i]
		i++
		k++
	}
	for j <= hi {
		tmp[k] = arr[j]
		j++
		k++
	}
	copy(arr[lo:hi+1], tmp)
}

// HeapSort 堆排序（原地）
func HeapSort(arr []int) {
	n := len(arr)
	for i := n/2 - 1; i >= 0; i-- {
		heapify(arr, n, i)
	}
	for i := n - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		heapify(arr, i, 0)
	}
}

func heapify(arr []int, n, i int) {
	largest := i
	left, right := 2*i+1, 2*i+2
	if left < n && arr[left] > arr[largest] {
		largest = left
	}
	if right < n && arr[right] > arr[largest] {
		largest = right
	}
	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		heapify(arr, n, largest)
	}
}
